File: string/stringCompare.ts


const getDerivedString = (string: string): string => {
    let updatedString: string[] = [];

    for (let char of string.split('')) {

        if (char === "#") {
            updatedString.pop();
            continue;
        }
        updatedString.push(char);
    }
    return updatedString.join("");
}

const compareStrings = (string1: string, string2: string): boolean => {

    const derivedString1 = getDerivedString(string1);
    const derivedString2 = getDerivedString(string2);
    console.log(derivedString1 === derivedString2);
    return false;
}

const optimalSolution = (string1: string, string2: string): boolean => {
    let p1 = string1.length - 1, p2 = string2.length - 1;
    while (p1 >= 0 || p2 >= 0) {
        if (string1[p1] === "#" || string2[p2] === "#") {
            if (string1[p1] === "#") {
                let backCount = 2;
                while (backCount > 0) {
                    p1--;
                    backCount--;
                    if (string1[p1] === "#") backCount += 2;
                }
            }

            if (string2[p2] === "#") {
                let backCount = 2;
                while (backCount > 0) {
                    p2--;
                    backCount--;
                    if (string2[p2] === "#") backCount += 2;
                }
            }
        } else {
            if (string1[p1] !== string2[p2]) {
                return false;
            }
            p1--;
            p2--;
        }
    }
    return true;
}


const isEqual = optimalSolution("ab#c#d", "ab#c#de#");
console.log(isEqual);

File: string/stringCompare.go
package main

import "fmt"

func optimalSolution(string1, string2 string) bool {

	p1 := len(string1) - 1
	p2 := len(string2) - 1

	for p1 >= 0 || p2 >= 0 {

		if string1[p1] == '#' || string2[p2] == '#' {
			if string1[p1] == '#' {
				backCount := 2

				for backCount > 0 {
					p1--
					backCount--

					if string1[p1] == '#' {
						backCount += 2
					}
				}
			}

			if string2[p2] == '#' {
				backCount := 2

				for backCount > 0 {
					p2--
					backCount--

					if string2[p2] == '#' {
						backCount += 2
					}
				}
			}
		} else {

			if string1[p1] != string2[p2] {
				return false
			}

			p1--
			p2--
		}
	}
	return true
}

func backspaceCompare(s string, t string) bool {
	p1 := len(s) - 1
	p2 := len(t) - 1

	for p1 >= 0 || p2 >= 0 {

		if s[p1] == '#' || t[p2] == '#' {
			if s[p1] == '#' {
				backCount := 2

				for backCount > 0 {
					p1--
					backCount--

					if p1 > 0 && s[p1] == '#' {
						backCount += 2
					}
				}
			}

			if t[p2] == '#' {
				backCount := 2

				for backCount > 0 {
					p2--
					backCount--

					if p2 > 0 && t[p2] == '#' {
						backCount += 2
					}
				}
			}
		} else {

			if s[p1] != t[p2] {
				return false
			}

			p1--
			p2--
		}
	}
	return true
}
func main() {

	isEqual := backspaceCompare("ab##", "c#d#")

	if isEqual {
		fmt.Println("The strings are equal")
	} else {
		fmt.Println("The strings are not equal")
	}
}
